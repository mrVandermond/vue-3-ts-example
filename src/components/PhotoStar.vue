<template>
  <div class="photo-star">
    <img
      :src="starUrl"
      alt="star-image"
    >
  </div>
</template>

<script lang="ts" setup>
import { computed } from 'vue';

const props = defineProps<{
  isActive: boolean;
}>();
const starUrl = computed((): string => {
  if (props.isActive) {
    return 'assets/star_active.png';
  }

  return 'assets/star_empty.png';
});
</script>

<style lang="scss">
.photo-star {
  width: 32px;
  height: 32px;
  border-radius: 50%;
  display: flex;
  align-items: center;
  justify-content: center;
}
</style>
