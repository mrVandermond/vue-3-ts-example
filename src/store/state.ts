import type { IModalConfig, IPhotoWithIsFavorite, ITitleConfig } from '@/types';

export interface IState {
  favoritePhotos: IPhotoWithIsFavorite[];
  modal: IModalConfig;
  title: ITitleConfig;
}

export const state = (): IState => ({
  favoritePhotos: [],
  modal: {
    isOpen: false,
    url: '',
  },
  title: {
    isShow: false,
    text: '',
    xCoord: 0,
    yCoord: 0,
  },
});
