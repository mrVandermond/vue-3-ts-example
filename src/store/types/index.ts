/* mutations */
export enum EMutations {
  ADD_TO_FAVORITES = 'ADD_TO_FAVORITES',
  REMOVE_FROM_FAVORITES = 'REMOVE_FROM_FAVORITES',
  SET_MODAL_CONFIG = 'SET_MODAL_CONFIG',
  SET_TITLE_CONFIG = 'SET_TITLE_CONFIG',
}
/* mutations */
/* getters */
export enum EGetters {
  IS_FAVORITE_PHOTO_BY_ID = 'IS_FAVORITE_PHOTO_BY_ID',
}
/* getters */
/* actions */
export enum EActions {
  GET_FAVORITE_PHOTOS_FROM_LOCAL_STORAGE = 'GET_FAVORITE_PHOTOS_FROM_LOCAL_STORAGE',
  SET_FAVORITE_PHOTOS_TO_LOCAL_STORAGE = 'SET_FAVORITE_PHOTOS_TO_LOCAL_STORAGE',
  FETCH_USERS = 'FETCH_USERS',
  FETCH_ALBUMS_BY_USER_ID = 'FETCH_ALBUMS_BY_USER_ID',
  FETCH_PHOTOS_BY_ALBUM_ID = 'FETCH_PHOTOS_BY_ALBUM_ID',
}
/* actions */

export const storeTypes = {
  EMutations,
  EGetters,
  EActions,
};
