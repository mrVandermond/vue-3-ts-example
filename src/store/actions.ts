import { storeTypes } from '@/store/types';
import { utlCst } from '@/utils';
import type { Mutations } from '@/store/mutations';
import type { ActionContext, ActionTree } from 'vuex';
import type { IState } from '@/store/state';
import type { IAlbum, IPhoto, IPhotoWithIsFavorite, IUser } from '@/types';
import * as api from '@/api';

type AugmentedActionContext = {
  commit<K extends keyof Mutations>(
    key: K,
    payload: Parameters<Mutations[K]>[1]
  ): ReturnType<Mutations[K]>;
} & Omit<ActionContext<IState, IState>, 'commit'>;

export interface Actions {
  [storeTypes.EActions.GET_FAVORITE_PHOTOS_FROM_LOCAL_STORAGE](
    ctx: AugmentedActionContext,
  ): void;
  [storeTypes.EActions.SET_FAVORITE_PHOTOS_TO_LOCAL_STORAGE](
    ctx: AugmentedActionContext,
  ): void;
  [storeTypes.EActions.FETCH_USERS](
    ctx: AugmentedActionContext,
  ): Promise<IUser[]>;
  [storeTypes.EActions.FETCH_ALBUMS_BY_USER_ID](
    ctx: AugmentedActionContext,
    payload: number,
  ): Promise<IAlbum[]>;
  [storeTypes.EActions.FETCH_PHOTOS_BY_ALBUM_ID](
    ctx: AugmentedActionContext,
    payload: number,
  ): Promise<IPhoto[]>;
}

export const actions: ActionTree<IState, IState> & Actions = {
  async [storeTypes.EActions.FETCH_USERS]() {
    const { data } = await api.fetchUsers();

    return data;
  },
  async [storeTypes.EActions.FETCH_ALBUMS_BY_USER_ID](ctx, userId) {
    const { data } = await api.fetchAlbumsByUserId(userId);

    return data;
  },
  async [storeTypes.EActions.FETCH_PHOTOS_BY_ALBUM_ID](ctx, albumId) {
    const { data } = await api.fetchPhotosByAlbumId(albumId);

    return data;
  },
  [storeTypes.EActions.GET_FAVORITE_PHOTOS_FROM_LOCAL_STORAGE]({ commit }) {
    const jsonFavoritePhotos = localStorage.getItem(utlCst.LOCAL_STORAGE_KEY);

    if (!jsonFavoritePhotos) return;

    const favoritePhotos = (JSON.parse(jsonFavoritePhotos) as IPhotoWithIsFavorite[]);

    commit(storeTypes.EMutations.ADD_TO_FAVORITES, favoritePhotos);
  },
  [storeTypes.EActions.SET_FAVORITE_PHOTOS_TO_LOCAL_STORAGE]({ state }) {
    const { favoritePhotos } = state;
    const jsonFavoritePhotos = JSON.stringify(favoritePhotos);

    localStorage.setItem(utlCst.LOCAL_STORAGE_KEY, jsonFavoritePhotos);
  },
};
