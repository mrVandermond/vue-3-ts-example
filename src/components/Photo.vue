<template>
  <div
    class="photo"
    :data-title="props.photo.title"
    @click="onClickOpenModal"
  >
    <PhotoStar
      :is-active="props.photo.isFavorite"
      class="photo-star"
      @click.stop="onClickStar"
    />

    <img
      :src="props.photo.thumbnailUrl"
      alt="photo-image"
    >
  </div>
</template>

<script lang="ts" setup>
import PhotoStar from '@/components/PhotoStar.vue';
import type { IPhotoWithIsFavorite } from '@/types';
import { customUseStore } from '@/store';
import { storeTypes } from '@/store/types';

const props = defineProps<{
  photo: IPhotoWithIsFavorite;
}>();
const emit = defineEmits<{
  (event: 'addedToFavorites', photo: IPhotoWithIsFavorite): void;
  (event: 'removeFromFavorites', photo: IPhotoWithIsFavorite): void;
}>();
const store = customUseStore();

const onClickStar = (): void => {
  if (props.photo.isFavorite) {
    emit('removeFromFavorites', props.photo);

    return;
  }

  emit('addedToFavorites', props.photo);
};
const onClickOpenModal = (): void => {
  store.commit(storeTypes.EMutations.SET_MODAL_CONFIG, {
    isOpen: true,
    url: props.photo.url,
  });
};
</script>

<style lang="scss">
.photo {
  overflow: hidden;
  border-radius: 4px;
  width: 150px;
  height: 150px;
  cursor: pointer;
  position: relative;
}
.photo-star {
  position: absolute;
  top: 8px;
  right: 8px;
}
</style>
