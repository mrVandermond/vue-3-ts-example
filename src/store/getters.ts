import type { GetterTree } from 'vuex';
import type { IState } from '@/store/state';
import { storeTypes } from '@/store/types';

export interface Getters {
  [storeTypes.EGetters.IS_FAVORITE_PHOTO_BY_ID](state: IState): (photoId: number) => boolean;
}

export const getters: GetterTree<IState, IState> & Getters = {
  [storeTypes.EGetters.IS_FAVORITE_PHOTO_BY_ID]: (state) => (photoId: number) => {
    const photo = state.favoritePhotos.find((item) => item.id === photoId);

    return !!photo;
  },
};
