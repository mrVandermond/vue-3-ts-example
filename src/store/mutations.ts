import { storeTypes } from '@/store/types';
import type { IModalConfig, IPhotoWithIsFavorite, ITitleConfig } from '@/types';
import type { IState } from '@/store/state';
import type { MutationTree } from 'vuex';
import { utlFuncs } from '@/utils';

export interface Mutations<S = IState> {
  [storeTypes.EMutations.ADD_TO_FAVORITES](state: S, payload: IPhotoWithIsFavorite[]): void;
  [storeTypes.EMutations.REMOVE_FROM_FAVORITES](state: S, payload: IPhotoWithIsFavorite): void;
  [storeTypes.EMutations.SET_MODAL_CONFIG](state: S, payload: IModalConfig): void;
  [storeTypes.EMutations.SET_TITLE_CONFIG](state: S, payload: ITitleConfig): void;
}

export const mutations: MutationTree<IState> & Mutations = {
  [storeTypes.EMutations.ADD_TO_FAVORITES](state, photos: IPhotoWithIsFavorite[]) {
    const photosWithoutExisting = photos.filter((photo) => (
      utlFuncs.getItemIndexBy(state.favoritePhotos, photo, 'id') === -1
    ));

    state.favoritePhotos.push(...photosWithoutExisting.map((photo) => ({ ...photo, isFavorite: true })));
  },
  [storeTypes.EMutations.REMOVE_FROM_FAVORITES](state, photo: IPhotoWithIsFavorite) {
    const photoIndex = utlFuncs.getItemIndexBy(state.favoritePhotos, photo, 'id');

    if (photoIndex === -1) return;

    state.favoritePhotos.splice(photoIndex, 1);
  },
  [storeTypes.EMutations.SET_MODAL_CONFIG](state, config) {
    state.modal.isOpen = config.isOpen;
    state.modal.url = config.url;
  },
  [storeTypes.EMutations.SET_TITLE_CONFIG](state, config) {
    state.title.isShow = config.isShow;
    state.title.text = config.text;
    state.title.xCoord = config.xCoord;
    state.title.yCoord = config.yCoord;
  },
};
