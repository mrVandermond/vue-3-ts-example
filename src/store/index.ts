import type { StoreOptions } from 'vuex';
import { createStore, useStore as baseUseStore } from 'vuex';
import type { InjectionKey } from 'vue';
import type { Store } from '@/store/types/overrides';
import type { IState } from './state';
import { state } from './state';
import { mutations } from './mutations';
import { getters } from './getters';
import { actions } from './actions';

export const key: InjectionKey<Store> = Symbol('store');

export function customUseStore(): Store {
  return baseUseStore(key);
}

export function customCreateStore(options: StoreOptions<IState>): Store {
  return createStore<IState>(options);
}

export const store = customCreateStore({
  state,
  mutations,
  getters,
  actions,
});
