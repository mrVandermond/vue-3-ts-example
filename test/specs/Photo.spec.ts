import Photo from '@/components/Photo.vue';
import PhotoStar from '@/components/PhotoStar.vue';

import type { IPhotoWithIsFavorite } from '@/types';
import type { Store } from '@/store/types/overrides';

import { customCreateStore, key } from '@/store';
import { storeTypes } from '@/store/types';
import { state } from '@/store/state';
import { mutations } from '@/store/mutations';
import { getters } from '@/store/getters';
import { actions } from '@/store/actions';

import type { TWrapperFactoryReturnType } from '../utils';
import { wrapperFactory } from '../utils';

const EXAMPLE_PHOTO: IPhotoWithIsFavorite = {
  id: 1,
  title: 'test_title',
  url: 'https://test_url',
  albumId: 1,
  thumbnailUrl: 'https://test_thumbnail_url',
  isFavorite: false,
};

describe('Photo component', () => {
  let store!: Store;
  const getWrapper = (photo = EXAMPLE_PHOTO): TWrapperFactoryReturnType => wrapperFactory(Photo, {
    props: {
      photo,
    },
    global: {
      plugins: [[store, key]],
    },
  });
  let commitSpy: jest.SpyInstance<ReturnType<Store['commit']>, jest.ArgsType<Store['commit']>>;

  beforeEach(() => {
    store = customCreateStore({
      state,
      mutations,
      getters,
      actions,
    });
    commitSpy = jest.spyOn(store, 'commit');
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('Рендеринг', () => {
    const wrapper = getWrapper();

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('Открытие модалки', async () => {
    const wrapper = getWrapper();
    const div = wrapper.find<HTMLDivElement>('.photo');

    await div.trigger('click');

    expect(commitSpy).toHaveBeenCalledWith(storeTypes.EMutations.SET_MODAL_CONFIG, expect.any(Object));
    expect(commitSpy.mock.calls[0][1]).toMatchSnapshot();
  });

  it('Клик по звездочке, когда фото не в избранном', async () => {
    const wrapper = getWrapper({
      ...EXAMPLE_PHOTO,
      isFavorite: false,
    });

    const photoStar = wrapper.findComponent(PhotoStar);

    await photoStar.trigger('click');

    expect(wrapper.emitted<IPhotoWithIsFavorite>().addedToFavorites).toBeTruthy();
    expect(commitSpy).not.toHaveBeenCalled();
  });

  it('Клик по звездочке, когда фото в избранном', async () => {
    const wrapper = getWrapper({
      ...EXAMPLE_PHOTO,
      isFavorite: true,
    });

    const photoStar = wrapper.findComponent(PhotoStar);

    await photoStar.trigger('click');

    expect(wrapper.emitted<IPhotoWithIsFavorite>().removeFromFavorites).toBeTruthy();
    expect(commitSpy).not.toHaveBeenCalled();
  });
});
