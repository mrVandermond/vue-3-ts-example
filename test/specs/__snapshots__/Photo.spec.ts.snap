// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Photo component > Открытие модалки 1`] = `
{
  "isOpen": true,
  "url": "https://test_url",
}
`;

exports[`Photo component > Рендеринг 1`] = `
"<div class=\\"photo\\" data-title=\\"test_title\\">
  <photo-star-stub isactive=\\"false\\" class=\\"photo-star\\"></photo-star-stub><img src=\\"https://test_thumbnail_url\\" alt=\\"photo-image\\">
</div>"
`;
